### ISSUE TITLE
questions about the implementation details

### ISSUE THREAD

**Issue by hrz2000** [mycfhs/DreamMix#1]
Congratulations on such a great job! I have some questions below and look forward to replying:
1. textual attribute substitution is only used in inference? is it right? What happens when used in training?
2. Is textual attribute substitution done only for pooled vectors of (bs,dim)? It seems to me that such an operation cannot be done for such a vector as (bs, seq, dim) because of the correspondence of the words.
3. by the way, we have a related work [FreeEdit: Mask-free Reference-based Image Editing with Multi-modal Instruction](https://arxiv.org/abs/2409.18071) that implements mask-free reference editing tasks, and we hope you can cite us in subsequent editions of the paper, if possibale.

**Comment by mycfhs** [comment]
Thank you for your attention to our work!


	1.	Our fine-tuning strategy will not change the embedding of words, with the aim of decoupling the target attributes during the inference stage. If the textual attribute substitution is applied to both training and inference simultaneously, it is equivalent to replacing "sks" with other word, which will have no effect.
	2.	We apply the textual attribute substitution to vectors and pooled vectors. For vectors, we operate at the positions corresponding to the input prompt in the sequence dimension, that is, [bs, 1:1 + promptlen, :]. You can refer to the code of orthogonal decomposition in utils/orthogonal_decomposition.py.
	3.	Of course, it's a great honor.


发自我的iPhone


------------------ Original ------------------
From: runzehe ***@***.***&gt;
Date: Wed,Nov 27,2024 3:21 PM
To: mycfhs/DreamMix ***@***.***&gt;
Cc: Subscribed ***@***.***&gt;
Subject: Re: [mycfhs/DreamMix] questions about the implementation details (Issue#1)





 
Congratulations on such a great job! I have some questions below and look forward to replying:
  
textual attribute substitution is only used in inference? is it right? What happens when used in training?
 
Is textual attribute substitution done only for pooled vectors of (bs,dim)? It seems to me that such an operation cannot be done for such a vector as (bs, seq, dim) because of the correspondence of the words.
 
by the way, we have a related work FreeEdit: Mask-free Reference-based Image Editing with Multi-modal Instruction that implements mask-free reference editing tasks, and we hope you can cite us in subsequent editions of the paper, if possibale.

**Comment by hrz2000** [comment]
thanks for your kindly reply! 

Your answer gave me a deeper understanding of this question. 

I still have a little doubt, for question 2, is there a theoretical basis for this token-by-token substitution approach or the implementation is found to be feasible empirically? Because I think that the semantics of each corresponding word of “raw vector” and “to_remove vector” are not the same as the semantic deletion we want to do (on a holistic level).

Especially, when the prompt_len of the two of them is different, it is difficult to define such behavior.

**Comment by mycfhs** [comment]
Considering that not only "sks" but also other words may fit unwanted features during the training process, we thus conducted a thorough decomposition. All vectors of the infer prompt were decomposed with respect to the to_remove vector respectively. Experiments show that this thorough decomposition will not affect the image quality.&nbsp;


I didn't quite get your point regarding the fact that the lengths of the two prompts you mentioned are not the same.&nbsp;Could you please explain it in more detail?

发自我的iPhone


------------------ Original ------------------
From: runzehe ***@***.***&gt;
Date: Wed,Nov 27,2024 5:06 PM
To: mycfhs/DreamMix ***@***.***&gt;
Cc: YiCheng Yang ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [mycfhs/DreamMix] questions about the implementation details(Issue #1)





 
thanks for your kindly reply!
 
Your answer gave me a deeper understanding of this question.
 
I still have a little doubt, for question 2, is there a theoretical basis for this token-by-token substitution approach or the implementation is found to be feasible empirically? Because I think that the semantics of each corresponding word of “raw vector” and “to_remove vector” are not the same as the semantic deletion we want to do (on a holistic level).
 
Especially, when the prompt_len of the two of them is different, it is difficult to define such behavior.

**Comment by hrz2000** [comment]
I'm sorry for not making it clear before, by double-checking your code, I realized that my previous idea was wrong：

raw="A sks glass teapot filled with green tea"
to_remove="clay brown"

my previous idea:
A-clay, sks-brown, ...

the actual implement:
A-clay, sks-clay, ...
A-brown, sks- brown, ...

Thank you again for your patience with me.

**Comment by mycfhs** [comment]
😊



发自我的iPhone


------------------ Original ------------------
From: runzehe ***@***.***&gt;
Date: Thu,Nov 28,2024 10:56 AM
To: mycfhs/DreamMix ***@***.***&gt;
Cc: YiCheng Yang ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [mycfhs/DreamMix] questions about the implementation details(Issue #1)





 
I'm sorry for not making it clear before, by double-checking your code, I realized that my previous idea was wrong：
 
raw="A sks glass teapot filled with green tea"
 to_remove="clay brown"
 
my previous idea:
 A-clay, sks-brown, ...
 
the actual implement:
 A-clay, sks-clay, ...
 A-brown, sks- brown, ...
 
Thank you again for your patience with me.
